File: Java/Concurrency/Concurrency_basics.java
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Executor;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;


class Pradeep implements Runnable{

    @Override
    public void run() {
        while(true){
            System.out.println("Pradeep is continously working");
            System.out.println("Pradeep Running this in a new Thread: " + Thread.currentThread().getName());
        }
    }
}

public class Concurrency_basics{

        public static void main(String[] args) {
            
            Pradeep p1 = new Pradeep();
            Pradeep p2 = new Pradeep();

            Thread t1 = new Thread(p1);
            Thread t2 = new Thread(p2);
            
            t1.start();
            t2.start();



            // ExecutorService pool = Executors.newFixedThreadPool(5); ///threadPool
            // Runnable task1 = ()->{
            //     for(int i = 0; i < 20; i++)
            //     System.out.println("pradeep");
            // };
            
            // pool.submit(task1);
            // pool.shutdown();
    /* 
        Runnable task1 = ()->{
            for(int i = 0; i < 20; i++)
            System.out.println("pradeep");
        };

        Runnable task2 = ()->{
            for(int i = 0; i < 20; i++)
            System.out.println("ishita");
        };

        Runnable task3 = ()->{
            for(int i = 0; i < 20; i++)
            System.out.println("Sahil");
        };

        Runnable [] tasks = new Runnable[3];
        tasks[0] = task1;
        tasks[1] = task2;
        tasks[2] = task3;

        List<Thread> threads = new ArrayList<>();
        
        for(Runnable task : tasks){
              Thread t1 = new Thread(task);
              threads.add(t1);  
              t1.start();
        }

        try {
            for(Thread t : threads){
                t.join();
            }
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        System.out.println("main thread");
    
    
        */
    }
  




}

File: Python/Concurrency/Concurrency_basics.py
import threading


class Pradeep:
    def run(self) -> None:
        while True:
            print("Pradeep is continously working")
            print(f"Pradeep Running this in a new Thread: {threading.current_thread().name}")


def main() -> None:
    p1 = Pradeep()
    p2 = Pradeep()

    t1 = threading.Thread(target=p1.run)
    t2 = threading.Thread(target=p2.run)

    t1.start()
    t2.start()

    # Alternative approaches considered:
    #
    # from concurrent.futures import ThreadPoolExecutor
    # def task1():
    #     for i in range(20):
    #         print("pradeep")
    # with ThreadPoolExecutor(max_workers=5) as pool:
    #     pool.submit(task1)
    #
    # def task1():
    #     for i in range(20):
    #         print("pradeep")
    # def task2():
    #     for i in range(20):
    #         print("ishita")
    # def task3():
    #     for i in range(20):
    #         print("Sahil")
    #
    # threads = []
    # for task in (task1, task2, task3):
    #     t = threading.Thread(target=task)
    #     threads.append(t)
    #     t.start()
    # for t in threads:
    #     t.join()
    # print("main thread")

    t1.join()
    t2.join()


if __name__ == "__main__":
    main()
